models/src/main/scala/coop/rchain/models/rholangn/Collection.scala

```diff
@@ -0,0 +1,111 @@
+package coop.rchain.models.rholangn
+
+import scala.collection.immutable.{TreeMap, TreeSet}
+
+/**
+  * Ordered collection of 0 or more processes.
+  * @param ps The sequence of any Rholang processes
+  * @param remainder Remainder of a list elements. This var used in matching (pattern of a head/tail pair)
+  */
+final class EListN(val ps: Seq[ParN], val remainder: Option[VarN]) extends CollectionN {
+  def :+(elem: ParN): EListN       = EListN(ps :+ elem, remainder)
+  def +:(elem: ParN): EListN       = EListN(elem +: ps, remainder)
+  def ++(elems: Seq[ParN]): EListN = EListN(ps ++ elems, None)
+  def ++(that: EListN): EListN     = EListN(ps ++ that.ps, None)
+}
```

Comment:
Reminder gives support to use `...` in the list construction and deconstruction e.g. `[1, 2, 3 ... rest]`. It's defined as optional variable.

But it should be preserved when concatenating lists.

Reply:
See T comment https://github.com/RhonixLabs/rhonix/pull/34#discussion_r1286760263
I did it like this - because that's how the previous implementation works:
https://github.com/rchain/rchain/blob/7497744de5363f76f9037437bd634c6e416fa8fc/rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala#L787
I don't know how to do it correctly - because it should only be used in pattern matching, and concatenation with the remainder should give an error. Maybe we should return an error in case of an existing remainder in operation with list?

Reply:
Ok, in the old implementation remainder of concat of two lists is empty. So I assume all improvements are for the future.

Reply:
To info for future, I dublicate it here:
I've tested some AST composing code with different remainders:
```scala
new x, y in { for(@z <- x) {y!([1, 2 ... z])}}
```
output: *Error*: Top level free variables are not allowed: z at 1:43.

```scala
new x, y in { for(_ <- x) {y!([1, 2 ... z])}}
```
output: *Error*: Top level free variables are not allowed: z at 1:43.

```scala
new x, y in { for( @{[z1, z2 ... rem1] ++ [z3, z4 ... rem2]} <-x) { y!(z1)} }
```
output: *Success*. Pretty print:
```scala
new x0, x1 in {
  for( @{([x3, x4...x2] ++ [x6, x7...x5])} <- x0 ) {
    x1!(x3)
  }
}
```
So, we can’t use reminder in reducer during evaluation. And there are not situation there you need to do concatenations with remainder. Only in pattern matching. So, it’s doesn’t matter what we will have in concat results. I think empty remainder after concatenations is a fine.